LFS-841: Refactor the pedigree data type

acrowthe opened the pull request:
- Add answerMetadata property to Answer component, enabling answer to be stored keyed, rather than in the value array if desired
- Move pedigree svg from sling value array to image property
- Change PedigreeQuestion to use defaultDisplayFormatter when not editing

acrowthe committed:
LFS-841: Refactor the pedigree data type
- Add answerMap property to Answer component, enabling answer to be
  stored keyed, rather than in the value array if desired
- Move pedigree svg from sling value array to image property
- Change PedigreeQuestion to use defaultDisplayFormatter when not editing

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -30,7 +30,7 @@ export const VALUE_POS = 1;
 // for form submission
 function Answer (props) {
-  let { answers, answerNodeType, existingAnswer, path, questionName, questionDefinition, valueType, onChangeNote, noteComponent, noteProps, onAddedAnswerPath, onDecidedOutputPath, sectionAnswersState } = props;
+  let { answers, answerMap, answerNodeType, existingAnswer, path, questionName, questionDefinition, valueType, onChangeNote, noteComponent, noteProps, onAddedAnswerPath, onDecidedOutputPath, sectionAnswersState } = props;
```
I would rename answerMap. One option would be answerMetadata. All suggestions to help convey that we're saving something related to the answer that is not its value but helps define it are welcome.

@sdumitriu I'm now wondering if it makes sense to save the ontology term label for vocab questions using this mechanism and improve search with it, at least until we get indexing fixed.

acrowthe replied:
Renamed to answerMetadata

marta- reviewed (approved):
All good other than the minor comment regarding a better name for `answerMap`.

acrowthe committed:
LFS-841: Refactor the pedigree data type
Rename answerMap to answerMetadata

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -82,6 +82,19 @@ function Answer (props) {
               );
           })}
+          {
+            answerMetadata ?
```
Prefer `answerMetadata && ...`

OrdiNeu reviewed (approved):
Looks good

acrowthe committed:
LFS-841: Refactor the pedigree data type
Slight refactor to improve readability

sashaandjic commented:
Not sure what the circled items are.
![image](https://user-images.githubusercontent.com/23243244/106215918-fd29c700-619f-11eb-8ad1-5f499ea0b976.png)

They appear for every type of pedigree ( to single empty node to multi-node pedigrees)

marta- commented:
> Not sure what the circled items are.

Good catch.

@acrowthe first of all the [JSON processor](https://github.com/ccmbioinfo/lfs/blob/f30157fb99063ccaa90431c7aba590b8d06a5eb2/modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/labels/PedigreeLabelProcessor.java#L57) needs to be updated to take the `displayedValue` from `image` instead of `value[1]`.

acrowthe committed:
LFS-841: Refactor the pedigree data type
- Change PedigreeLabelProcessor to use image property
- Fix duplicate image saved by pedigree question

sashaandjic reviewed (approved):
tested - all good
![image](https://user-images.githubusercontent.com/23243244/106539550-362da880-64cc-11eb-9878-ae1045a825c4.png)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Answer.jsx:
```diff
@@ -82,6 +82,18 @@ function Answer (props) {
+                    type="hidden"
+                    name={`${answerPath}/${key}`}
+                    key={value === undefined ? index + (answers ? answers.length : 0) : value}
+                    value={value}></input>
```
I think support for multivalued entries should be added as well. If value is an array, add an input for each subvalue.
